### ISSUE TITLE
codex review

### ISSUE THREAD

**Issue by bitranox** [rsionnach/sloppylint#17]
I reviewed the **latest published release, sloppylint 0.4.0**, by inspecting the PyPI sdist and wheel contents, plus the upstream release notes. ([[PyPI](https://pypi.org/project/sloppylint/)][1])

## High level assessment

The project has a solid architecture for what it is trying to do. A small CLI and config layer feed a detector pipeline. The detector runs AST patterns, then line based regex patterns, then file level analyzers, then optional cross file duplicate analysis. The patterns are sensibly grouped by “axis” and severity, and the output model is consistent.

The biggest quality gaps are in. Version and naming consistency, path glob semantics, and a couple of places where environment dependence can create noisy results.

## What improved in the new version

From the release notes and the shipped CLI/config. The new version adds include filtering, integrates Ruff, adds pre commit hooks, adds Dependabot, adds pip audit, and makes mypy blocking in CI. ([[GitHub](https://github.com/rsionnach/sloppylint/releases)][2])

That is all directionally great. It reduces contributor friction and makes regressions less likely.

### Good design choices

* **Pattern system is extensible.** `BasePattern` plus `RegexPattern` and `ASTPattern` is a clean abstraction. It keeps pattern authors from touching the detector orchestration.
* **False positive controls are built in.** Patterns that can trigger inside strings or comments use a shared helper approach, and the detector computes multiline string line ranges once per file and injects it into patterns before scanning.
* **Separation of concerns is mostly correct.** AST walking lives in `ASTAnalyzer`. Heavier analyses are in dedicated analyzers. Reporting and scoring are separate modules.

### Where the design will start to creak

* **Detector is becoming a “god orchestrator”.** It now owns file IO, AST parsing, pattern dispatch, analyzer dispatch, filtering, and cross file pass control. It is still readable, but you are close to the point where small feature additions will cause lots of branching.
* **Two glob engines in play.** `Path.match` plus a custom `fnmatch` helper means users will see weird edge cases. Especially across OS path separators and absolute vs relative matching.

### 1. Version reporting is wrong

In 0.4.0, `pyproject.toml` says version **0.4.0**, but `src/sloppy/__init__.py` still exports `__version__ = "0.1.0"`. That means `sloppylint --version` is very likely incorrect if it reads from `__version__`. ([[PyPI](https://pypi.org/project/sloppylint/)][1])

Fix. Set `__version__` from the build metadata or generate it during build. With Hatch, common options are.

* Single source of truth in `pyproject.toml`, then import via `importlib.metadata.version("sloppylint")`.
* Or Hatch version file generation, if you want a constant in code.

This is the most important “polish bug” because it breaks trust immediately.

### 2. CLI program name mismatch

The installed command is `sloppylint` on [PyPI](https://pypi.org/project/sloppylint/) and via `project.scripts`. ([PyPI][1])
But the argparse `prog` string is still set to `"sloppy"`, which leaks into help and error output. This is cosmetic, but it is also user facing.

Fix. Set `prog="sloppylint"` or let argparse infer it.

### 3. Glob matching semantics can surprise users

* Ignore patterns use `Path.match(pattern)` only. Users tend to expect gitignore style matching, or at least `**` style recursion.
* Include patterns use both `Path.match` and `_match_glob`, but `_match_glob` mixes raw string paths with `fnmatch`. That can break on Windows path separators, and can behave differently for absolute paths.

Fix. Pick one matching model and make it predictable.

* Compute a project root. Probably the current working directory unless explicitly set.
* Match against relative POSIX style paths. `path.relative_to(root).as_posix()`.
* Use `fnmatch` on the POSIX string, and document that patterns are matched against relative paths.
* Apply the same engine to include and ignore.

### 4. Environment dependent import validation can be noisy

`import_validator.module_exists` checks the current Python environment using `importlib.util.find_spec`. That is useful for catching hallucinated packages, but it will also flag real third party imports if the user has not installed dependencies in the environment running sloppylint.

There is some mitigation in the wrong stdlib and known confusion table. Still, the tool is implicitly mixing “static analysis of a repo” with “what is installed right now”.

Fix options.

* Add a mode that only validates stdlib and obvious wrong patterns. Default that on in CI contexts.
* Add support for reading dependencies from `pyproject.toml` and `requirements.txt` so it can treat declared deps as “allowed even if not installed”.
* Add a config allowlist for module roots.

### 5. Defensive exception handling could hide real errors

`module_exists` returns True on unexpected exceptions to avoid false positives. That is understandable, but it can mask genuine errors where spec lookup is failing consistently.

Fix. Consider logging debug output when `--verbose` is enabled, and track “skipped validations” as a low severity informational issue.

### Strengths

* The new helper based approach to “ignore strings and inline comments” is exactly the kind of shared infrastructure you want in a linter, because it prevents every pattern from re inventing partial parsing.
* Several patterns show thoughtful exclusions, which reduces obvious noise.

### Risks

* The regex pattern set is large. As it grows, you will need a stronger internal structure.

  * Pattern IDs must stay stable.
  * Messages need consistency and clarity.
  * Tests must pin false positive behavior tightly, or you will regress constantly.
* Multi line string handling is still a known limitation per release notes. The helper improves it, but it is hard to solve perfectly with line based matching alone. ([[GitHub](https://github.com/rsionnach/sloppylint/releases)][2])

Recommendation. For the noisiest patterns, consider moving from regex scanning to token based scanning using Python’s `tokenize` module. It gives you exact comment and string boundaries without maintaining your own state machine.

### Unused imports

The approach is conventional AST walking plus a “string annotation” heuristic. It will be correct in many cases, but it will miss things like.

* `typing.TYPE_CHECKING` blocks.
* `getattr(module, "Name")` usage.
* Re exports via `__all__`.

This is fine if you position it as “AI slop smell” rather than strict correctness. Still, you may want a config toggle to disable it by default, or to downgrade severity.

### Dead code

Dead code detection is one of the hardest problems to do well without full semantic analysis. The current approach is a reasonable heuristic, but users will see false positives in frameworks, plugin registries, decorators, and dynamic call paths.

Recommendation. Make sure you document typical suppression patterns. Offer `# sloppylint: disable=dead_code` style inline disables at the line or symbol level.

### Duplicate detection

Cross file duplicates can be valuable, but it is also expensive. If a user points this at a large monorepo, runtime could spike.

Recommendation.

* Add a file count or byte budget limit for the cross file pass by default.
* Or make it opt in via `--duplicates` or config, unless you already have strong performance.

## Tooling and project hygiene

The repo is moving in the right direction. Ruff, pre commit, Dependabot, pip audit, strict mypy in CI. All good. ([[GitHub](https://github.com/rsionnach/sloppylint/releases)][2])

Two things I would still add.

* A small “golden corpus” of fixture files that represent realistic AI slop patterns and tricky false positive cases. Run the tool against it in CI and snapshot the JSON output.
* Property based tests for helper logic like string and comment detection. That code tends to be brittle.

## Packaging and distribution

* PyPI release is current and clearly documented. ([[PyPI](https://pypi.org/project/sloppylint/)][1])
* Wheel contents are clean. The sdist contains extra internal directories like `.beads`, but the wheel does not. If you care about sdist cleanliness, exclude those via your build config.

## Priority fix list

If you want the highest impact improvements first.

1. Fix the version mismatch. `__version__` must match 0.4.0. ([[PyPI](https://pypi.org/project/sloppylint/)][1])
2. Fix argparse `prog` naming to match `sloppylint`. ([[PyPI](https://pypi.org/project/sloppylint/)][1])
3. Unify include and ignore matching with a single documented glob engine.
4. Reduce environment coupling in import validation, or make it configurable.
5. Add corpus based regression tests for false positives.

If you want, I can also produce a concrete patch plan with exact code changes for items 1 to 3, including how to make glob matching consistent across platforms, and how to make version reporting single source of truth.

[1]: https://pypi.org/project/sloppylint/ "sloppylint · PyPI"
[2]: https://github.com/rsionnach/sloppylint/releases "Releases · rsionnach/sloppylint · GitHub"

**Comment by rsionnach** [comment]
Thanks again for your input. Absolutely, feel free to open a PR with your proposed changes! 👍

**Comment by bitranox** [comment]
phu, have enough on the plate myself ... 
I considered sloppylint as additional linter for my project  https://github.com/bitranox/glintefy

**Comment by rsionnach** [comment]
Nice! I will 💯 review this in the next few days with your proposals. This project is a secondary focus for me. My main focus at the moment is https://github.com/rsionnach/nthlayer.

**Comment by rsionnach** [comment]
## Addressed items 1 and 2 from the priority fix list:

### 1. Version mismatch - FIXED
Changed `src/sloppy/__init__.py` to use `importlib.metadata.version()` instead of hardcoded version:
```python
try:
    from importlib.metadata import version
    __version__ = version("sloppylint")
except Exception:
    __version__ = "0.0.0"  # Fallback for development
```
This ensures version is always read from pyproject.toml (single source of truth).

### 2. CLI program name - FIXED
Changed argparse `prog="sloppy"` to `prog="sloppylint"` to match the installed command name.

---

**Remaining items from this review:**
- Item 3 (glob matching) - Still open
- Item 4 (environment-dependent imports) - Still open  
- Item 5 (corpus tests) - Partially addressed (we have corpus tests)

**Items resolved by issue #19 refactoring:**
- Unused imports analyzer - Removed (use ruff F401)
- Dead code analyzer - Removed (use vulture)
- Duplicate detection - Removed (use CPD/pylint)

**Comment by rsionnach** [comment]
## Addressed items 3 and 4 from the priority fix list:

### 3. Glob matching - FIXED
Unified glob matching with a single consistent engine:

- All path matching now uses POSIX-style relative paths (cross-platform consistent)
- Single `_match_pattern()` method handles both include and ignore patterns
- Proper `**` recursive matching using regex conversion
- Removed the inconsistent mix of `Path.match()` and `fnmatch`

**How it works now:**
```python
# Paths are converted to relative POSIX paths from root
rel_path = path.resolve().relative_to(root).as_posix()

# Patterns like 'src/**/*.py' properly match 'src/a/b/c.py'
# Same engine for both --include and --ignore
```

### 4. Environment-dependent import validation - FIXED
Added `--strict-imports` flag (default: off) to control behavior:

**Default (non-strict):** Only flags KNOWN AI hallucination patterns:
- `utils`, `helpers`, `common` (generic names AI invents)
- Known wrong-module imports (e.g., `from requests import JSONResponse`)
- Does NOT check if third-party packages are installed (avoids false positives)

**Strict mode (`--strict-imports`):** Also validates against installed packages:
- Flags any import that can't be found in the current environment
- Useful when running in the project's virtualenv
- May cause false positives for dependencies not yet installed

**Configuration:**
```toml
# pyproject.toml
[tool.sloppy]
strict-imports = true  # Enable strict checking
```

Or via CLI:
```bash
sloppylint --strict-imports src/
```

---

**All priority items from this review are now complete:**
1. Version mismatch - FIXED (uses importlib.metadata)
2. CLI prog name - FIXED (now 'sloppylint')
3. Glob matching - FIXED (unified engine)
4. Environment-dependent imports - FIXED (strict-imports flag)
5. Corpus tests - Already have corpus-based tests

All 34 tests pass.
